src/lib.rs

```diff
@@ -689,33 +737,31 @@ decl_module! {
                             LegStatus::ExecutionToBeSkipped(receipt.signer.clone(), receipt.receipt_uid)
                         );
                     } else {
-                        match T::Asset::unsafe_increase_custody_allowance(
-                            did,
-                            leg_details.asset,
-                            did,
-                            SettlementDID.as_id(),
-                            leg_details.amount
-                        ) {
-                            Ok(_) => <InstructionLegStatus<T>>::insert(instruction_id, leg_id, LegStatus::ExecutionPending),
-                            Err(_) => return Rollback(Err(Error::<T>::FailedToTakeCustodialOwnership))
+                        if T::Portfolio::lock_tokens(&leg_details.from, &leg_details.asset, &leg_details.amount).is_err() {
+                            // rustc fails to infer return type of `with_transaction` if you use ?/map_err here
+                            return Err(DispatchError::from(Error::<T>::FailedToLockTokens));
                         }
+                        <InstructionLegStatus<T>>::insert(instruction_id, leg_id, LegStatus::ExecutionPending);
+
                     }
                 }
-                Commit(Ok(()))
+                Ok(())
             })?;
 
             // Update storage
-            let auths_pending = Self::instruction_auths_pending(instruction_id).saturating_sub(1);
-            <UserAuths>::insert(did, instruction_id, AuthorizationStatus::Authorized);
-            <AuthsReceived>::insert(instruction_id, did, AuthorizationStatus::Authorized);
+            let auths_pending = Self::instruction_auths_pending(instruction_id).saturating_sub(u64::try_from(portfolios.len()).unwrap_or_default());
+            for portfolio in portfolios {
```

Comment:
can we use `zip` to iterate both portfolio and receipt in one for loop?

Reply:
But why? portfolios and receipts are different things.

Reply:
It is more about saving one loop.